### PR TITLE
Add dynamic adding / removing stream providers functionality.

### PR DESCRIPTION
**Pull request by JingqiaoFu**
This change will add the functionality of dynamically adding / removing stream providers on silo at runtime, so that each silo doesn't have to be restarted to pick up newly added stream providers in the config file.

this feature can programmatically manipulate configuration objects in the silo instance. We do not address the issue where this new functionality may cause differences between the configuration object and the config files. Any dynamic change made at runtime will NOT be recorded back into config files. Therefore, if a silo is restarted, the changes made last time will be lost.  

User experience:  

1. User creates a stream provider configuration object, which contains a list of specifications of all the stream providers that needs to be loaded. 

2. User calls member function UpdateStreamProviders(streamProviderConfig) of management grain or silo itself, passes in the configuration object as the parameter. The stream providers in target silos will be updated based on the differences between their loaded stream providers and the list of providers in the object: If a provider in the configuration object already exists in a silo, it will be kept as is; if a provider in the configuration object does not exist in a silo, it will be loaded and started; if a provider that exists in a silo but is not in the configuration object, it will be stopped and removed from the silo.

### CODE REVIEW

**Comment by jdom** [comment]
@dotnet-bot test this please
reason: removed dependency on MSTest since the PR was created

File: src/Orleans/Streams/SimpleMessageStream/SimpleMessageStreamProvider.cs [hunk 1]
```diff
@@ -8,12 +8,14 @@ namespace Orleans.Providers.Streams.SimpleMessageStream
     public class SimpleMessageStreamProvider : IInternalStreamProvider
     {
         public string                       Name { get; private set; }
+        private ProviderState               State { get; set; }
```
Comment (jason-bragg) [inline src/Orleans/Streams/SimpleMessageStream/SimpleMessageStreamProvider.cs]: Why track this separately, why not just ask the stateManager for the state?

File: src/Orleans/Providers/IOrleansProvider.cs [hunk 2]
```diff
@@ -160,4 +161,23 @@ public static TimeSpan GetTimeSpanProperty(this IProviderConfiguration config, s
             return config.Properties.TryGetValue(key, out s) ? TimeSpan.Parse(s) : settingDefault;
         }
     }
+
+    /// <summary>
+    /// Exception thrown whenever a provider has failed to be initialized.
+    /// </summary>
+    [Serializable]
+    public class ProviderInitializationException : OrleansException
```
Comment (gabikliot) [inline src/Orleans/Providers/IOrleansProvider.cs]: please put all the new exception in one file, better in the same file where all the other  OrleansExceptions are now.

Comment (jason-bragg) [inline src/Orleans/Providers/IOrleansProvider.cs]: We added it here because it's an addition to the public surface of the provider model.  On initialization failure, providers are expected to throw this exception.

File: src/OrleansRuntime/Silo/Silo.cs [hunk 3]
```diff
@@ -487,13 +491,13 @@ private void DoStart()
             if (logger.IsVerbose) { logger.Verbose("Storage provider manager created successfully."); }
 
             // Load and init stream providers before silo becomes active
-            var siloStreamProviderManager = (StreamProviderManager) grainRuntime.StreamProviderManager;
+            var streamProviderManager = (StreamProviderManager)grainRuntime.StreamProviderManager;
```
Comment (gabikliot) [inline src/OrleansRuntime/Silo/Silo.cs]: why do you need to rename that? Lets keep unrelated churn low. If you want to refactor, unrelated to the change, please submit it is a separate PR.

File: test/TesterInternal/StreamingTests/DynamicStreamProviderConfigurationTests.cs [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+                DefaultStreamProviderSettings.Add(PersistentStreamProviderConfig.STREAM_PUBSUB_TYPE, StreamPubSubType.ImplicitOnly.ToString());
+
+                return new TestCluster(options);
+            }
+        }
+
+        public async void Dispose()
+        {
+            await RemoveAllProviders();
+            fixture.DefaultStreamProviderSettings.Remove(FailureInjectionStreamProvider.FailureInjectionModeString);
```
Comment (jdom) [inline test/TesterInternal/StreamingTests/DynamicStreamProviderConfigurationTests.cs]: This is odd. Please do no update the fixture itself, as it is mutable state across tests and would break as soon as we add some test parallelization (#1863)

File: test/TesterInternal/StreamingTests/DynamicStreamProviderConfigurationTests.cs [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+        [Fact, TestCategory("Functional"), TestCategory("Streaming"), TestCategory("Providers")]
+        public async Task DynamicStreamProviderConfigurationTests_AddProvidersAndThrowExceptionOnInitialization()
+        {
+            //Making sure the initial provider list is empty.
+            IEnumerable<String> providerNames = fixture.HostedCluster.Primary.Silo.TestHook.GetStreamProviderNames();
+            Assert.Equal(providerNames.Count(), 0);
+
+            bool exceptionThrown = false;
+            fixture.DefaultStreamProviderSettings.Remove(FailureInjectionStreamProvider.FailureInjectionModeString);
+            fixture.DefaultStreamProviderSettings.Add(FailureInjectionStreamProvider.FailureInjectionModeString,
```
Comment (jdom) [inline test/TesterInternal/StreamingTests/DynamicStreamProviderConfigurationTests.cs]: Same here, please never update the shared fixture on a test by test basis

File: src/Orleans/Streams/SimpleMessageStream/SimpleMessageStreamProvider.cs (hunk 1 above)
Comment (JingqiaoFu) [inline src/Orleans/Streams/SimpleMessageStream/SimpleMessageStreamProvider.cs]: Good catch. This is obsolete code that I missed to delete.

File: test/TesterInternal/StreamingTests/DynamicStreamProviderConfigurationTests.cs [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+                DefaultStreamProviderSettings.Add(PersistentStreamProviderConfig.STREAM_PUBSUB_TYPE, StreamPubSubType.ImplicitOnly.ToString());
+
+                return new TestCluster(options);
+            }
+        }
+
+        public async void Dispose()
+        {
+            await RemoveAllProviders();
+            //fixture.DefaultStreamProviderSettings.Remove(FailureInjectionStreamProvider.FailureInjectionModeString);
```
Comment (jdom) [inline test/TesterInternal/StreamingTests/DynamicStreamProviderConfigurationTests.cs]: latest tests look good, just remove this comment

**Comment by jason-bragg** [comment]
Looks good.
Please rebase and squash, and I'll merge it.

Also did you sign the contributors license agreement?  PR is still marked cla-required.

File: test/Tester/TestStreamProviders/FailureInjectionStreamProvider.cs [hunk 7]
```diff
@@ -0,0 +1,69 @@
+
+using System;
+using System.Collections.Generic;
+using System.Management.Automation.Language;
```
Comment (jdom) [inline test/Tester/TestStreamProviders/FailureInjectionStreamProvider.cs]: be careful with including namespaces that are not really used. This is currently harmless in .NET Framework, but it's breaking when we multi target to .NET Core (which we are doing in other branches). This is why we recently removed all superfluous using statements in #1985. We should be careful in new code not to add more.
